### PR TITLE
Da/transaction service

### PR DESCRIPTION
**Pull request by axelrod-blox**

### CODE REVIEW

File: src/app/common/config/config.ts [hunk 1]
```diff
@@ -238,7 +238,7 @@ const DEFAULT_PAGINATION = {
   page: 1,
   pages: 1,
   total: 0,
-  per_page: 8,
+  per_page: 10,
```
Comment (axelrod-blox) [inline src/app/common/config/config.ts:241]: why is 10? we don't have this requirement

File: src/app/common/stores/applications/Distribution/DistributionTestnet.store.ts [hunk 2]
```diff
@@ -86,9 +85,9 @@ class DistributionTestnetStore extends BaseStore {
     await this.cleanState();
     // @ts-ignore
     const merkleTreeAddresses = Object.keys(merkleTree.claims);
-    const walletStore: WalletStore = this.getStore('Wallet');
+    const accountAddress = store.getState().walletState.accountAddress;
     merkleTreeAddresses.forEach((address: string) => {
-      if (address.toLowerCase() === walletStore.accountAddress.toLowerCase()) {
+      if (address.toLowerCase() === accountAddress.toLowerCase()) {
```
Comment (axelrod-blox) [inline src/app/common/stores/applications/Distribution/DistributionTestnet.store.ts:90]: we already has function in utils **equalsAddresses** please use it

File: src/app/components/applications/SSV/MyAccount/components/Validator/EditFlow/OperatorsRecipt/OperatorsReceipt.tsx [hunk 3]
```diff
@@ -61,7 +61,7 @@ const OperatorsReceipt = (props: Props) => {
 
   const newOperatorsFee = operators.reduce(
     (previousValue: number, currentValue: IOperator) => {
-      if (currentValue.ownerAddress !== walletStore.accountAddress) {
+      if (currentValue.ownerAddress !== accountAddress) {
```
Comment (axelrod-blox) [inline src/app/components/applications/SSV/MyAccount/components/Validator/EditFlow/OperatorsRecipt/OperatorsReceipt.tsx:64]: also use **equalsAddresses**

File: src/app/common/stores/applications/SsvWeb/Validator.store.ts [hunk 4]
```diff
@@ -18,12 +16,13 @@ import { setIsLoading, setIsShowTxPendingPopup, setTxHash } from '~app/redux/app
...
 import { getClusterData, getClusterHash, getSortedOperatorsIds } from '~root/services/cluster.service';
 import { getLiquidationCollateralPerValidator, getValidator } from '~root/services/validator.service';
-import { getEventByTxHash } from '~root/services/contractEvent.service';
 import { translations } from '~app/common/config';
 import { getOwnerNonce } from '~root/services/account.service';
 import { checkEntityChangedInAccount, delay } from '~root/services/utils.service';
 import { SingleCluster, RegisterValidator } from '~app/model/processes.model';
 import { setMessageAndSeverity } from '~app/redux/notifications.slice';
+import { transactionExecutor, TransactionMethod } from '~root/services/transaction.service';
+import { dispatch } from 'jest-circus/build/state';
```
Comment (IlyaVi) [inline src/app/common/stores/applications/SsvWeb/Validator.store.ts:25]: remove this import

File: src/app/common/stores/applications/SsvWeb/Validator.store.ts [hunk 5]
```diff
@@ -152,38 +150,31 @@ class ValidatorStore extends BaseStore {
...
-        }
-        if (isContractWallet) {
-          resolve(true);
-        }
-        const receipt = await tx.wait();
-        if (receipt.blockHash) {
-          await executeAfterEvent(async () => !await getValidator(publicKey), async () => this.myAccountStore.refreshOperatorsAndClusters(), delay);
-          resolve(true);
-        } else {
+        const sortedOperatorIds = getSortedOperatorsIds(operatorIds);
```
Comment (IlyaVi) [inline src/app/common/stores/applications/SsvWeb/Validator.store.ts:158]: simplify - remove external try catch

File: src/app/common/stores/applications/SsvWeb/Validator.store.ts [hunk 6]
```diff
@@ -152,38 +150,31 @@ class ValidatorStore extends BaseStore {
...
+        const clusterData = await getClusterData(getClusterHash(operatorIds, accountAddress), ssvStore.liquidationCollateralPeriod, ssvStore.minimumLiquidationCollateral);
```
Comment (IlyaVi) [inline src/app/common/stores/applications/SsvWeb/Validator.store.ts:159]: getClusterData - unnecessary try catch in the function
simplify the function itself and check for proper error value handling

File: src/app/common/stores/applications/SsvWeb/Validator.store.ts [hunk 7]
```diff
@@ -152,38 +150,31 @@ class ValidatorStore extends BaseStore {
...
   async removeValidator(publicKey: string, operatorIds: number[]): Promise<boolean> {
-    const contract = getContractByName(EContractName.SETTER);
-    store.dispatch(setIsLoading(true));
     const isContractWallet = store.getState().walletState.isContractWallet;
     const accountAddress = store.getState().walletState.accountAddress;
     const ssvStore: SsvStore = this.getStore('SSV');
-    const sortedOperatorIds = getSortedOperatorsIds(operatorIds);
-    const clusterData = await getClusterData(getClusterHash(operatorIds, accountAddress), ssvStore.liquidationCollateralPeriod, ssvStore.minimumLiquidationCollateral);
     return new Promise(async (resolve) => {
       try {
```
Comment (IlyaVi) [inline src/app/common/stores/applications/SsvWeb/Validator.store.ts:157]: no need to wrap in Promise and do manual resolve or reject

File: src/app/components/applications/SSV/MyAccount/components/Validator/EditFlow/OperatorsRecipt/OperatorsReceipt.tsx [hunk 8]
```diff
@@ -80,16 +80,16 @@ const OperatorsReceipt = (props: Props) => {
...
-    dispatch(setIsLoading(true));
-    const response = await validatorStore.updateValidator();
-    if (response) {
-      navigate(config.routes.SSV.MY_ACCOUNT.CLUSTER.VALIDATOR_UPDATE.SUCCESS);
-      setTimeout(() => {
-        navigate(config.routes.SSV.MY_ACCOUNT.CLUSTER.ROOT);
-      }, 10000);
-    }
-  };
+  // const updateValidator = async () => {
```
Comment (IlyaVi) [inline src/app/components/applications/SSV/MyAccount/components/Validator/EditFlow/OperatorsRecipt/OperatorsReceipt.tsx]: remove

File: src/app/components/applications/SSV/MyAccount/components/Validator/EditFlow/OperatorsRecipt/OperatorsReceipt.tsx [hunk 9]
```diff
@@ -175,7 +175,7 @@ const OperatorsReceipt = (props: Props) => {
         <Button
           // @ts-ignore
           text={'Update Operators'}
-          onClick={updateValidator}
+          // onClick={updateValidator}
```
Comment (IlyaVi) [inline src/app/components/applications/SSV/MyAccount/components/Validator/EditFlow/OperatorsRecipt/OperatorsReceipt.tsx]: remove this Button component

File: src/app/common/stores/applications/SsvWeb/Validator.store.ts [hunk 10]
```diff
@@ -152,38 +150,31 @@ class ValidatorStore extends BaseStore {
...
+        if (!payload) {
           resolve(false);
+          return;
         }
+        await transactionExecutor({
+          methodType: TransactionMethod.RemoveValidator,
+          payload,
+          contract,
+          isContractWallet: isContractWallet,
+          callbackAfterExecution: async () => await this.myAccountStore.refreshOperatorsAndClusters(),
```
Comment (IlyaVi) [inline src/app/common/stores/applications/SsvWeb/Validator.store.ts:171]: pass pointer to function 
`callbackAfterExecution: this.myAccountStore.refreshOperatorsAndClusters`

File: src/services/transaction.service.ts [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+import { store } from '~app/store';
+import { setIsLoading, setIsShowTxPendingPopup, setTxHash } from '~app/redux/appState.slice';
+import { setMessageAndSeverity } from '~app/redux/notifications.slice';
+import { translations } from '~app/common/config';
+import { executeAfterEventV2 } from '~root/services/events.service';
+import { getEventByTxHash } from '~root/services/contractEvent.service';
+import { delay } from '~root/services/utils.service';
+import { getValidator } from '~root/services/validator.service';
+
+export enum TransactionMethod {
```
Comment (IlyaVi) [inline src/services/transaction.service.ts:11]: move to src/app/enums

File: src/services/transaction.service.ts [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+export const transactionExecutor = async ({ methodType, contract, payload, isContractWallet, callbackAfterExecution }: {
+  methodType: TransactionMethod,
+  contract: Contract,
+  payload: any,
+  isContractWallet: boolean,
+  callbackAfterExecution: Function,
+}) => {
+  try {
+    let tx = await contractMethods[methodType](contract, payload);
```
Comment (IlyaVi) [inline src/services/transaction.service.ts:50]: you can simplify it by passing pointer to contract method
something like this 
`transactionExecutor({ contractMethod: contract.registerValidator, ...`

File: src/services/transaction.service.ts [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    }
+    return false;
+  } catch (e: any) {
+    store.dispatch(setMessageAndSeverity({
+      message: e.message || translations.DEFAULT.DEFAULT_ERROR_MESSAGE,
+      severity: 'error',
+    }));
+    return false;
+  } finally {
+    await callbackAfterExecution();
```
Comment (IlyaVi) [inline src/services/transaction.service.ts:90]: are you sure to call callback even if there is an error?

File: src/services/transaction.service.ts [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+const finishTransactionExecutor = {
+  [TransactionMethod.RegisterValidator]: async (hash: string) => await executeAfterEventV2(async () => !!await getEventByTxHash(hash), delay),
+  [TransactionMethod.BulkRegisterValidator]: async (hash: string) => await executeAfterEventV2(async () => !!await getEventByTxHash(hash), delay),
+  [TransactionMethod.RemoveValidator]: async (publicKey: string) => await executeAfterEventV2(async () => !await getValidator(publicKey, true), delay),
+  [TransactionMethod.BulkRemoveValidator]: async (publicKey: string) => await executeAfterEventV2(async () => !await getValidator(publicKey, true), delay),
+  [TransactionMethod.ReactivateCluster]: async (hash: string) => await executeAfterEventV2(async () => !!await getEventByTxHash(hash), delay),
+  [TransactionMethod.UpdateValidator]: async (hash: string) => await executeAfterEventV2(async () => !!await getEventByTxHash(hash), delay),
+};
+
+export const transactionExecutor = async ({ methodType, contract, payload, isContractWallet, callbackAfterExecution }: {
```
Comment (IlyaVi) [inline src/services/transaction.service.ts:42]: send payload as array, ...Object.keys is expensive operation

Comment (axelrod-blox) [reply src/services/transaction.service.ts:42]: +

**Review by IlyaVi — APPROVED** [review]

**Pull request merged**